### PR TITLE
feat: add Safety Error for non-unit parameter test

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes https://github.com/flix/flix/issues/4287

### CODE REVIEW

**Comment by jaschdoc** [comment]
Hmm... I'm not sure what to do with the tests in https://github.com/flix/flix/blob/master/main/test/flix/Test.Exp.Fixpoint.flix

**Comment by magnus-madsen** [comment]
Remove the `@Test`  from those that take arguments.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 1]
```diff
@@ -64,7 +64,7 @@ object Safety {
     */
   private def visitDefs(root: Root)(implicit flix: Flix): List[CompilationMessage] = {
     root.defs.flatMap {
-      case (_, defn) => visitDef(defn, root)
+      case (_, defn) => visitDef(defn, root) ::: visitTestEntryPoint(defn, root)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:67]: Is there a reason this is not part of visitDef`?

File: main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala [hunk 2]
```diff
@@ -553,4 +553,29 @@ object SafetyError {
          |""".stripMargin
     )
   }
+
+  case class IllegalTestParameters(loc: SourceLocation) extends SafetyError {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala:557]: We always have scaladoc for errors

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 3]
```diff
@@ -78,6 +78,30 @@ object Safety {
...
+
+    def isUnitType(fparam: FormalParam): Boolean = fparam.tpe.typeConstructor.contains(TypeConstructor.Unit)
+
+    def hasParameters(d: Def, loc: SourceLocation): Boolean = d.spec.fparams match {
+      case fparam :: Nil if isUnitType(fparam) => false
+      case Nil => throw InternalCompilerException("Unexpected empty formal parameter list near", loc)
+      case _ => true
+    }
+
+    root.defs.get(def0.sym).map {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:95]: Why is this lookup needed? Don't we already have the def in hand?

File: main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala (hunk 2 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala:557]: Oh yes I forgot that

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala (hunk 1 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Safety.scala:67]: I thought it was clumsy in that function, but I experimented with it being there. I can put it back there if you want.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala (hunk 3 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Safety.scala:95]: Can we reach the spec from there?

**Comment by magnus-madsen** [comment]
> > Remove the `@Test`  from those that take arguments.
> 
> I'm not sure how I feel about just removing tests. Are we going to try to add them back later?

Remove the annotation, not the code itself :)

The test is really that it type checks, not that it runs.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala (hunk 3 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Safety.scala:95]: Ok yeah I see :facepalm:

File: main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala [hunk 4]
```diff
@@ -553,4 +553,35 @@ object SafetyError {
...
+         |>> Test entry point must not have parameters.
+
+         |${code(loc, "Parameter for test function occurs here.")}
+         |
+         |""".stripMargin
+    }
+
+    def explain(formatter: Formatter): Option[String] = Some(
+      s"""
+         |A test function must not have parameters.
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala:578]: Would this be clearer with an example?

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala (hunk 4 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/errors/SafetyError.scala:578]: Yes

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 5]
```diff
@@ -75,7 +75,30 @@ object Safety {
...
       case (acc, e) => acc.markRigid(e)
     }
-    visitExp(def0.impl.exp, renv, root)
+    visitTestEntryPoint(def0, root) ::: visitExp(def0.impl.exp, renv, root)
+  }
+
+  /**
+    * Checks that if `def0` is a test entry point that it is well-behaved.
+    */
+  private def visitTestEntryPoint(def0: Def, root: Root): List[CompilationMessage] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:84]: Lets immediately do this test: `def0.spec.ann.isTest`

if (def0.spec.ann.isTest) {... } else {Nil}

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 6]
```diff
@@ -75,7 +75,30 @@ object Safety {
...
+
+    val hasParameters = def0.spec.fparams match {
+      case fparam :: Nil if isUnitType(fparam) => false
+      case Nil => throw InternalCompilerException("Unexpected empty formal parameter list near", def0.spec.loc)
+      case _ => true
+    }
+
+    if (isTest && hasParameters) {
+      val fparam :: _ = def0.spec.fparams
+      val err = SafetyError.IllegalTestParameters(fparam.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:97]: This is overkill. Lets grab the loc from def0.sym.loc.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 7]
```diff
@@ -75,7 +75,30 @@ object Safety {
...
+
+  /**
+    * Checks that if `def0` is a test entry point that it is well-behaved.
+    */
+  private def visitTestEntryPoint(def0: Def, root: Root): List[CompilationMessage] = {
+    def isUnitType(fparam: FormalParam): Boolean = fparam.tpe.typeConstructor.contains(TypeConstructor.Unit)
+
+    val isTest = def0.spec.ann.isTest
+
+    val hasParameters = def0.spec.fparams match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:89]: I would flip this "hasUnitParameter" and drop the exception case.

**Comment by jaschdoc** [comment]
Ready for review again

**Pull request merged**